### DISCUSSION TITLE
Dynamic version doesn't apply to wheel

### DISCUSSION THREAD

**Discussion by roysmith** [pypa/setuptools#3752] [Q&A] [answered]
I'm using pip 22.3.1, python 3.9.13, MacOS.  Using this pyproject.toml file:

```
[build-system]
requires = ["setuptools"]
build-backend = "setuptools.build_meta"

[project]
name = "dyk_tools"
# dynamic = ["version"]
version = "0.0.1"
dependencies = [
  "Flask",
  "mwoauth",
  "mwparserfromhell",
  "pytest-mock",
  "pytest-socket",
  "pywikibot",
  "sh",
  "ujson",
  "WTforms",
]

[project.scripts]
get-states = "dyk_tools.scripts:get_states.main"
analyze-article = "dyk_tools.scripts:analyze_article.main"
find-approved_templates = "dyk_tools.scripts:find_approved_templates.main"

[tool.setuptools]
packages = ["dyk_tools", "dyk_tools.scripts"]


[tool.setuptools.dynamic]
version = {file = ["VERSION"]}
```

and "0.0.1" in my VERSION file, python -m build gets me whl and tar files with the 0.0.1 version:

```
(dyk-tools) Roys-MacBook-Pro [dev2] ls -l dist
total 40
-rw-r--r--  1 roy  staff  11066 Jan  6 13:36 dyk_tools-0.0.1-py3-none-any.whl
-rw-r--r--  1 roy  staff   7536 Jan  6 13:36 dyk_tools-0.0.1.tar.gz
```
So far, so good.  But if I turn on the dynamic version

```
dynamic = ["version"]
# version = "0.0.1"
```

my whl file has 0.0.0 for the version:

```
(dyk-tools) Roys-MacBook-Pro [dev2] ls -l dist
total 40
-rw-r--r--  1 roy  staff  11065 Jan  6 13:41 dyk_tools-0.0.0-py3-none-any.whl
-rw-r--r--  1 roy  staff   7541 Jan  6 13:41 dyk_tools-0.0.1.tar.gz
```

**Comment by webknjaz** [comment]
How did you run the build?

**Reply by roysmith** [reply]
I ran "python -m build"

**Reply by abravalheri** [reply]
Hi @roysmith could you please list the contents of `dyk_tools-0.0.1.tar.gz`?

`python -m build` will generate the wheel from the sdist (not PWD). In order to be used for the wheel, the `VERSION` file needs to be part of the sdist, this is controlled either by using an automatic plugin like `setuptools-scm` or by manually setting `MANIFEST.in` (as discussed in the [docs](https://setuptools.pypa.io/en/latest/userguide/miscellaneous.html)).

**Comment by roysmith** [comment, marked as answer]
Answer: https://github.com/pypa/setuptools/discussions/3752#discussioncomment-4615331

Yup, that was it.  No VERSION in the tar file.   I updated my pyproject.toml:

```
[build-system]
requires = ["setuptools>=45", "setuptools_scm[toml]>=6.2"]
build-backend = "setuptools.build_meta"

...

[tool.setuptools_scm]
```

and reran the build.  Still no VERSION in the tar :-(  Then I actually committed the VERSION file to git, and _now_ I've got VERSION in the tar file and the whl file is properly named:

```
-rw-r--r--  1 roy  staff  13828 Jan  6 15:00 dyk_tools-0.0.1-py3-none-any.whl
-rw-r--r--  1 roy  staff  12332 Jan  6 15:00 dyk_tools-0.0.1.tar.gz
```

Thanks for your help!

**Reply by webknjaz** [reply]
FWIW setuptools-scm is mainly useful to generate the version from the Git tag+distance. If you are planning to hardcode the version in a file, make sure to understand the implications. Also, setuptools-scm can generate a Python file and add it to the sdist without having to commit it FYI.

**Reply by roysmith** [reply]
Actually, generating the version from a git tag is exactly where I'm heading.  Reading it from the VERSION file was just a temporary step while I figured out how all this new-fangled packaging stuff worked :-)
